pkg/operator2/status.go

```diff
@@ -1,75 +1,113 @@
 package operator2
 
 import (
-	"strings"
+	"k8s.io/apimachinery/pkg/util/sets"
 
 	operatorv1 "github.com/openshift/api/operator/v1"
 	"github.com/openshift/library-go/pkg/operator/v1helpers"
 )
 
-func handleDegraded(operatorConfig *operatorv1.Authentication, prefix string, err error) {
-	handleDegradedWithReason(operatorConfig, prefix, "Error", err)
+// Each prefix included in this set will have its condition defaulted
+// to false if not explicitly set to true.
+var degradedConditionPrefixes = sets.NewString(
+	"IdentityProviderConfig",
+	"OAuthClients",
+	"OperatorSync",
+	"RouteHealth",
+	"RouteStatus",
+	"WellKnownEndpoint",
+)
+
+type authConditions struct {
+	conditions  []operatorv1.OperatorCondition
+	hasDegraded bool
+}
+
+func newAuthConditions() *authConditions {
+	return &authConditions{
+		conditions: []operatorv1.OperatorCondition{},
+	}
 }
 
-func handleDegradedWithReason(operatorConfig *operatorv1.Authentication, prefix, reason string, err error) {
-	if err != nil {
-		v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions,
-			operatorv1.OperatorCondition{
-				Type:    prefix + operatorv1.OperatorStatusTypeDegraded,
-				Status:  operatorv1.ConditionTrue,
-				Reason:  reason,
-				Message: err.Error(),
+// all returns the set of conditions representing the complete status
+// of the auth operator. If a condition was not set, it will be
+// returned in its default state (degraded -> false, progressing ->
+// false, available -> true).
+func (c *authConditions) all() []operatorv1.OperatorCondition {
```

Comment:
`func (c *authConditions) allUpdates() []v1helpers.UpdateStatusFunc {` so that in `Sync()` we can just do:

```golang
if _, _, err := v1helpers.UpdateStatus(c.authOperatorConfigClient, conditions.allUpdates()...)
```
instead of the unnecessary conditions copying magic we do now

Reply:
I purposely didn't chooses the approach you are suggesting because the current call to `UpdateStatus` needs to do more than update conditions.

Reply:
See: https://github.com/openshift/cluster-authentication-operator/blob/fix-ingress-state-flapping/pkg/operator2/operator.go#L206

Reply:
If the correct link is https://github.com/marun/cluster-authentication-operator/blob/fix-ingress-state-flapping/pkg/operator2/operator.go#L206 then I still don't see what the problem is. You set the `OperatorSyncDegraded` to true or false and then can still call `if _, _, err := v1helpers.UpdateStatus(c.authOperatorConfigClient, conditions.allUpdates()...)`

Am I missing something?

Reply:
Sorry, my comment was made prior to our conversation on Friday and is no longer relevant. Resolving...

Reply:
I think there was a bit of a misunderstanding - I was fine with the `all()` funcs, I just wanted you to form it into `allFuncs()` so tha we can do a simple `UpdateStatus()` call with it as I pointed out in https://github.com/openshift/cluster-authentication-operator/pull/269#discussion_r410163230 which would remove our custom update status logic and just reused https://github.com/openshift/library-go/blob/ccc4bb27d0320da04e5d2af6be09f030440bb152/pkg/operator/v1helpers/helpers.go#L184-L190